File: native/android/app/src/main/kotlin/art/lazying/auspice/screens/Astrology.kt
package art.lazying.auspice.screens

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.*
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.scale
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.drawIntoCanvas
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import art.lazying.auspice.*
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.roundToInt
import kotlin.math.sin
import kotlinx.serialization.json.buildJsonObject

@Composable
fun AstrologyScreen(navController: NavController) {
    var chart by remember { mutableStateOf<NatalChart?>(null) }
    var report by remember { mutableStateOf<TransitReport?>(null) }
    var error by remember { mutableStateOf<String?>(null) }
    var editing by remember { mutableStateOf(false) }
    val profile = Profiles.profile

    LaunchedEffect(profile) {
        if (!profile.isComplete) { chart = null; return@LaunchedEffect }
        runCatching { Engines.evaluateAs<NatalChart>("astrology.chart", profile.engineInput()) }
            .onSuccess { chart = it; error = null }
            .onFailure { error = it.message }
        report = runCatching {
            Engines.evaluateAs<TransitReport>(
                "astrology.transits",
                buildJsonObject { put("birth", profile.engineInput()) }
            )
        }.getOrNull()
    }

    if (editing) BirthFormDialog(profile, { editing = false }) { Profiles.save(it) }

    ScreenScaffold(
        navController,
        eyebrow = t("practice.astrology"),
        title = t("astro.title"),
        tagline = t("astro.tagline")
    ) {
        BirthSummary(profile) { editing = true }

        error?.let { Panel(title = t("common.notComputed")) { Text(it, style = Type.serif(16), color = Palette.inkSoft) } }

        chart?.let { c ->
            Panel {
                ChartWheel(c)
                Row(
                    Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceEvenly
                ) {
                    AngleLabel(t("astro.ascendant"), c.ascendant)
                    AngleLabel(t("astro.midheaven"), c.midheaven)
                }
            }

            Panel(title = t("astro.placements")) {
                c.placements.forEachIndexed { index, placement ->
                    Row(
                        Modifier.fillMaxWidth().padding(vertical = 5.dp),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(12.dp)
                    ) {
                        Text(
                            Zodiac.bodySymbols[placement.body] ?: "",
                            style = Type.display(21), color = Palette.gold, modifier = Modifier.width(26.dp)
                        )
                        Text(
                            placement.body,
                            style = Type.sans(15, FontWeight.SemiBold), color = Palette.ink,
                            modifier = Modifier.width(80.dp)
                        )
                        Text(degreeText(placement.longitude), style = Type.serif(16), color = Palette.inkSoft)
                        if (placement.retrograde) Text("℞", style = Type.display(15), color = Palette.rose)
                        Spacer(Modifier.weight(1f))
                        Text("${t("astro.house")} ${placement.house}", style = Type.sans(12), color = Palette.inkMute)
                    }
                    if (index < c.placements.lastIndex) HorizontalDivider(color = Palette.line)
                }
            }

            if (c.aspects.isNotEmpty()) {
                Panel(title = t("astro.aspects")) {
                    c.aspects.forEach { aspect ->
                        Row(
                            Modifier.fillMaxWidth().padding(vertical = 4.dp),
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(10.dp)
                        ) {
                            Text(
                                "${Zodiac.bodySymbols[aspect.a] ?: ""} ${aspectGlyph(aspect.type)} ${Zodiac.bodySymbols[aspect.b] ?: ""}",
                                style = Type.display(18), color = aspectColour(aspect.type)
                            )
                            Text(
                                "${aspect.a} ${aspect.type} ${aspect.b}",
                                style = Type.serif(15), color = Palette.inkSoft
                            )
                            Spacer(Modifier.weight(1f))
                            Text(String.format("%.1f°", aspect.orb), style = Type.sans(12), color = Palette.inkMute)
                        }
                    }
                }
            }

            report?.takeIf { it.transits.isNotEmpty() }?.let { r ->
                Panel(title = t("astro.transits")) {
                    r.transits.take(12).forEach { transit ->
                        Row(
                            Modifier.fillMaxWidth().padding(vertical = 3.dp),
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(10.dp)
                        ) {
                            Text(
                                "${Zodiac.bodySymbols[transit.transiting] ?: ""} ${aspectGlyph(transit.type)} ${Zodiac.bodySymbols[transit.natal] ?: ""}",
                                style = Type.display(18), color = aspectColour(transit.type)
                            )
                            Text(
                                "transiting ${transit.transiting} ${transit.type} natal ${transit.natal}",
                                style = Type.serif(15), color = Palette.inkSoft
                            )
                        }
                    }
                }
            }

            Panel(title = t("common.method")) {
                Text(
                    "Positions come from the astronomy engine for the exact instant of birth in UTC, with " +
                        "whole-sign houses counted from the ascendant. The moon stands at " +
                        "${(c.moonPhase * 100).roundToInt()}% of its cycle.",
                    style = Type.serif(16), color = Palette.inkSoft
                )
            }
        }
    }
}

@Composable
private fun AngleLabel(label: String, degrees: Double) {
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Text(
            label.uppercase(),
            style = Type.sans(10, FontWeight.Bold).copy(letterSpacing = 1.6.sp),
            color = Palette.inkMute
        )
        Text(degreeText(degrees), style = Type.display(17), color = Palette.gold)
    }
}

/** Signs on the rim, whole-sign houses from the ascendant, aspects across the middle. */
@Composable
private fun ChartWheel(chart: NatalChart) {
    var appeared by remember { mutableStateOf(false) }
    LaunchedEffect(chart) { appeared = true }
    val enter by animateFloatAsState(if (appeared) 1f else 0f, tween(1100), label = "wheel")

    Canvas(
        Modifier
            .fillMaxWidth()
            .aspectRatio(1f)
            .alpha(enter)
            .scale(0.92f + 0.08f * enter)
    ) {
        val centre = Offset(size.width / 2, size.height / 2)
        val extent = minOf(size.width, size.height)
        val outer = extent * 0.47f
        val inner = extent * 0.33f
        val aspectRadius = extent * 0.30f

        fun at(radius: Float, longitude: Double): Offset {
            val angle = (180 - longitude) * PI / 180
            return Offset(centre.x + (cos(angle) * radius).toFloat(), centre.y - (sin(angle) * radius).toFloat())
        }

        drawCircle(Palette.goldLine, outer, centre, style = Stroke(1f))
        drawCircle(Palette.line, inner, centre, style = Stroke(1f))
        for (index in 0 until 12) {
            val longitude = index * 30.0
            drawLine(Palette.line, at(inner, longitude), at(outer, longitude), 1f)
            val house = chart.ascendantSign * 30.0 + index * 30.0
            drawLine(Palette.gold.copy(alpha = 0.25f), at(aspectRadius, house), at(inner, house), 1f)
        }
        chart.aspects.forEach { aspect ->
            val a = chart.placements.firstOrNull { it.body == aspect.a } ?: return@forEach
            val b = chart.placements.firstOrNull { it.body == aspect.b } ?: return@forEach
            drawLine(
                when (aspect.type) {
                    "trine", "sextile" -> Color(0x8C7BC47F)
                    "square", "opposition" -> Palette.rose.copy(alpha = 0.5f)
                    else -> Palette.gold.copy(alpha = 0.5f)
                },
                at(aspectRadius, a.longitude), at(aspectRadius, b.longitude), 1f
            )
        }
        drawLine(Palette.gold, at(aspectRadius, chart.ascendant), at(outer, chart.ascendant), 2f)

        drawIntoCanvas { canvas ->
            val paint = android.graphics.Paint().apply {
                color = android.graphics.Color.argb(128, 244, 239, 228)
                textAlign = android.graphics.Paint.Align.CENTER
                textSize = extent * 0.05f
                isAntiAlias = true
            }
            for (index in 0 until 12) {
                val point = at((outer + inner) / 2, index * 30.0 + 15)
                canvas.nativeCanvas.drawText(Zodiac.signs[index].third, point.x, point.y + paint.textSize / 3, paint)
            }
            paint.color = android.graphics.Color.rgb(217, 180, 90)
            paint.textSize = extent * 0.062f
            chart.placements.forEach { placement ->
                val point = at(inner * 0.86f, placement.longitude)
                canvas.nativeCanvas.drawText(
                    Zodiac.bodySymbols[placement.body] ?: "",
                    point.x, point.y + paint.textSize / 3, paint
                )
            }
        }
    }
}

private fun degreeText(longitude: Double): String {
    val sign = ((longitude / 30).toInt()) % 12
    val within = longitude - sign * 30
    val degrees = within.toInt()
    val minutes = ((within - degrees) * 60).toInt()
    return String.format("%d°%02d′ %s %s", degrees, minutes, Zodiac.signs[sign].third, Zodiac.signs[sign].second)
}

private fun aspectGlyph(type: String) = when (type) {
    "conjunction" -> "☌"
    "sextile" -> "⚹"
    "square" -> "□"
    "trine" -> "△"
    else -> "☍"
}

private fun aspectColour(type: String) = when (type) {
    "trine", "sextile" -> Color(0xFF7BC47F)
    "square", "opposition" -> Palette.rose
    else -> Palette.gold
}

File: native/ios/Auspice/Screens/Astrology.swift
import SwiftUI

struct AstrologyScreen: View {
    @State private var store = ProfileStore.shared
    @SavedPractice("astrology.chart") private var chart: NatalChart? = nil
    @SavedPractice("astrology.report") private var report: TransitReport? = nil
    @State private var error: String?
    @SavedPractice("astrology.profile") private var computedProfile: BirthProfile? = nil
    @State private var editing = false
    @State private var wheelIn = false

    var body: some View {
        ScreenScaffold(
            eyebrow: t("practice.astrology"),
            title: t("astro.title"),
            tagline: t("astro.tagline")
        ) {
            BirthSummary(profile: store.profile) { editing = true }

            if let error {
                Panel(title: t("common.notComputed")) {
                    Text(error).font(Typeface.serif(16)).foregroundStyle(Palette.inkSoft)
                }
            }

            if let chart {
                Panel {
                    ChartWheel(chart: chart, appeared: wheelIn)
                        .aspectRatio(1, contentMode: .fit)
                        .frame(maxWidth: .infinity)
                        .onAppear {
                            withAnimation(.easeOut(duration: 1.1)) { wheelIn = true }
                        }
                    HStack(spacing: 16) {
                        angle(t("astro.ascendant"), chart.ascendant)
                        angle(t("astro.midheaven"), chart.midheaven)
                    }
                    .frame(maxWidth: .infinity)
                }

                Panel(title: t("astro.placements")) {
                    ForEach(chart.placements) { placement in
                        placementRow(placement)
                        if placement.id != chart.placements.last?.id {
                            Divider().overlay(Palette.line)
                        }
                    }
                }

                if !chart.aspects.isEmpty {
                    Panel(title: t("astro.aspects")) {
                        ForEach(chart.aspects) { aspect in
                            HStack(spacing: 10) {
                                Text(Zodiac.bodySymbols[aspect.a] ?? aspect.a)
                                    .font(Typeface.display(19))
                                    .foregroundStyle(Palette.ink)
                                Text(aspectGlyph(aspect.type))
                                    .font(Typeface.display(17))
                                    .foregroundStyle(aspectColour(aspect.type))
                                Text(Zodiac.bodySymbols[aspect.b] ?? aspect.b)
                                    .font(Typeface.display(19))
                                    .foregroundStyle(Palette.ink)
                                Text("\(l(aspect.a)) · \(l(aspect.type)) · \(l(aspect.b))")
                                    .font(Typeface.serif(15))
                                    .foregroundStyle(Palette.inkSoft)
                                Spacer(minLength: 0)
                                Text(String(format: "%.1f°", aspect.orb))
                                    .font(Typeface.sans(12))
                                    .foregroundStyle(Palette.inkMute)
                            }
                            .padding(.vertical, 4)
                        }
                    }
                }

                if let report, !report.transits.isEmpty {
                    Panel(title: t("astro.transits")) {
                        ForEach(report.transits.prefix(12)) { transit in
                            HStack(spacing: 10) {
                                Text("\(Zodiac.bodySymbols[transit.transiting] ?? "") \(aspectGlyph(transit.type)) \(Zodiac.bodySymbols[transit.natal] ?? "")")
                                    .font(Typeface.display(18))
                                    .foregroundStyle(aspectColour(transit.type))
                                Text(lf("Transiting {0} · {1} · natal {2}", transit.transiting, transit.type, transit.natal))
                                    .font(Typeface.serif(15))
                                    .foregroundStyle(Palette.inkSoft)
                                Spacer(minLength: 0)
                                Text(String(format: "%.1f°", transit.orb))
                                    .font(Typeface.sans(12))
                                    .foregroundStyle(Palette.inkMute)
                            }
                            .padding(.vertical, 3)
                        }
                    }
                }

                Panel(title: t("common.method")) {
                    Text(lf("Positions are calculated at the birth instant in UTC, using whole-sign houses from the ascendant. Lunar cycle: {0}%.", String(Int((chart.moonPhase * 100).rounded()))))
                        .font(Typeface.serif(16))
                        .foregroundStyle(Palette.inkSoft)
                        .fixedSize(horizontal: false, vertical: true)
                }
                ExplainReading(result: chart)
            }
        }
        .sheet(isPresented: $editing) {
            BirthForm(profile: $store.profile) { compute() }
        }
        .onAppear { if chart == nil || computedProfile != store.profile { compute() } }
    }

    private func angle(_ label: String, _ degrees: Double) -> some View {
        VStack(spacing: 2) {
            Text(label.uppercased())
                .font(Typeface.sans(10, weight: .bold))
                .tracking(1.6)
                .foregroundStyle(Palette.inkMute)
            Text(degreeText(degrees))
                .font(Typeface.display(17))
                .foregroundStyle(Palette.gold)
        }
    }

    private func placementRow(_ placement: Placement) -> some View {
        HStack(spacing: 12) {
            Text(Zodiac.bodySymbols[placement.body] ?? "")
                .font(Typeface.display(21))
                .foregroundStyle(Palette.gold)
                .frame(width: 26)
            Text(l(placement.body))
                .font(Typeface.sans(15, weight: .semibold))
                .foregroundStyle(Palette.ink)
                .frame(width: 78, alignment: .leading)
            Text(degreeText(placement.longitude))
                .font(Typeface.serif(16))
                .foregroundStyle(Palette.inkSoft)
            if placement.retrograde {
                Text("℞").font(Typeface.display(15)).foregroundStyle(Palette.rose)
            }
            Spacer(minLength: 0)
            Text("\(t("astro.house")) \(placement.house)")
                .font(Typeface.sans(12))
                .foregroundStyle(Palette.inkMute)
        }
        .padding(.vertical, 5)
    }

    private func degreeText(_ longitude: Double) -> String {
        let sign = Int(longitude / 30) % 12
        let within = longitude - Double(sign) * 30
        let degrees = Int(within)
        let minutes = Int((within - Double(degrees)) * 60)
        return String(format: "%d°%02d′ %@ %@", degrees, minutes, Zodiac.signs[sign].symbol, l(Zodiac.signs[sign].en))
    }

    private func aspectGlyph(_ type: String) -> String {
        switch type {
        case "conjunction": "☌"
        case "sextile": "⚹"
        case "square": "□"
        case "trine": "△"
        default: "☍"
        }
    }

    private func aspectColour(_ type: String) -> Color {
        switch type {
        case "trine", "sextile": Color(hex: 0x7BC47F)
        case "square", "opposition": Palette.rose
        default: Palette.gold
        }
    }

    private func compute() {
        guard store.profile.isComplete else { chart = nil; return }
        do {
            let input = store.profile.engineInput
            chart = try Engines.shared.evaluate("astrology.chart", input, as: NatalChart.self)
            report = try? Engines.shared.evaluate("astrology.transits", ["birth": input], as: TransitReport.self)
            error = nil
            computedProfile = store.profile
        } catch {
            self.error = l("This reading could not be computed. Please try again.")
        }
    }
}

/// The chart wheel: twelve signs, the houses from the ascendant, the aspect
/// lines across the middle, and the bodies on the rim.
struct ChartWheel: View {
    let chart: NatalChart
    let appeared: Bool

    var body: some View {
        GeometryReader { geometry in
            let size = min(geometry.size.width, geometry.size.height)
            let centre = CGPoint(x: geometry.size.width / 2, y: geometry.size.height / 2)
            let outer = size * 0.47
            let inner = size * 0.33
            let aspectRadius = size * 0.30

            ZStack {
                Canvas { context, _ in
                    // The rim, and the twelve sign divisions.
                    context.stroke(
                        Path(ellipseIn: CGRect(x: centre.x - outer, y: centre.y - outer, width: outer * 2, height: outer * 2)),
                        with: .color(Palette.goldLine),
                        lineWidth: 1
                    )
                    context.stroke(
                        Path(ellipseIn: CGRect(x: centre.x - inner, y: centre.y - inner, width: inner * 2, height: inner * 2)),
                        with: .color(Palette.line),
                        lineWidth: 1
                    )
                    for index in 0..<12 {
                        let angle = self.angle(for: Double(index) * 30)
                        var path = Path()
                        path.move(to: point(centre, inner, angle))
                        path.addLine(to: point(centre, outer, angle))
                        context.stroke(path, with: .color(Palette.line), lineWidth: 1)
                    }
                    // The houses, counted whole-sign from the ascendant.
                    for index in 0..<12 {
                        let angle = self.angle(for: Double(chart.ascendantSign) * 30 + Double(index) * 30)
                        var path = Path()
                        path.move(to: point(centre, aspectRadius, angle))
                        path.addLine(to: point(centre, inner, angle))
                        context.stroke(path, with: .color(Palette.gold.opacity(0.25)), lineWidth: 1)
                    }
                    // Aspect lines, drawn inside.
                    for aspect in chart.aspects {
                        guard let a = chart.placements.first(where: { $0.body == aspect.a }),
                              let b = chart.placements.first(where: { $0.body == aspect.b }) else { continue }
                        var path = Path()
                        path.move(to: point(centre, aspectRadius, self.angle(for: a.longitude)))
                        path.addLine(to: point(centre, aspectRadius, self.angle(for: b.longitude)))
                        context.stroke(path, with: .color(colour(aspect.type)), lineWidth: 1)
                    }
                    // The ascendant, marked in gold across the left.
                    var ascendant = Path()
                    ascendant.move(to: point(centre, aspectRadius, self.angle(for: chart.ascendant)))
                    ascendant.addLine(to: point(centre, outer, self.angle(for: chart.ascendant)))
                    context.stroke(ascendant, with: .color(Palette.gold), lineWidth: 2)
                }

                // Sign glyphs on the rim.
                ForEach(0..<12, id: \.self) { index in
                    Text(Zodiac.signs[index].symbol)
                        .font(.system(size: size * 0.05))
                        .foregroundStyle(Palette.inkMute)
                        .position(point(centre, (outer + inner) / 2, angle(for: Double(index) * 30 + 15)))
                }

                // The bodies, just inside the rim.
                ForEach(chart.placements) { placement in
                    Text(Zodiac.bodySymbols[placement.body] ?? "")
                        .font(.system(size: size * 0.062))
                        .foregroundStyle(Palette.gold)
                        .position(point(centre, inner * 0.86, angle(for: placement.longitude)))
                }
            }
            .rotationEffect(.degrees(appeared ? 0 : -18))
            .opacity(appeared ? 1 : 0)
            .scaleEffect(appeared ? 1 : 0.92)
        }
    }

    /// Zero degrees Aries to the left, counting anticlockwise, as charts are drawn.
    private func angle(for longitude: Double) -> Double {
        (180 - longitude) * .pi / 180
    }

    private func point(_ centre: CGPoint, _ radius: CGFloat, _ angle: Double) -> CGPoint {
        CGPoint(
            x: centre.x + CGFloat(Foundation.cos(angle)) * radius,
            y: centre.y - CGFloat(Foundation.sin(angle)) * radius
        )
    }

    private func colour(_ type: String) -> Color {
        switch type {
        case "trine", "sextile": Color(hex: 0x7BC47F).opacity(0.55)
        case "square", "opposition": Palette.rose.opacity(0.5)
        default: Palette.gold.opacity(0.5)
        }
    }
}
